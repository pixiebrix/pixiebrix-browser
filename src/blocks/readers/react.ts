/*
 * Copyright (C) 2020 Pixie Brix, LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import { createSendScriptMessage } from "@/messaging/chrome";
import { READ_REACT_COMPONENT } from "@/messaging/constants";
import { ReaderOutput } from "@/core";
import { registerFactory } from "@/blocks/readers/factory";

export interface ReactConfig {
  type: "react";
  selector: string;
  traverseUp?: number;
  rootProp?: string;
  waitMillis?: number;
}

export const withReactComponent = createSendScriptMessage<ReaderOutput>(
  READ_REACT_COMPONENT
);

async function doRead(reader: ReactConfig): Promise<ReaderOutput> {
  const { selector, traverseUp = 0, waitMillis = 1000, rootProp } = reader;
  return await withReactComponent({
    selector,
    traverseUp,
    waitMillis,
    rootProp,
  });
}

registerFactory("react", doRead);
