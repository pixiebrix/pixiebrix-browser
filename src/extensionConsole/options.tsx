/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import "@/vendors/theme/app/app.scss";
import "@/vendors/overrides.scss";
import "@/utils/layout.scss";
import "./options.scss";

import "@/extensionContext";
import "@/development/darkMode";

import { render } from "react-dom";
import React from "react";
import App from "@/extensionConsole/App";
import initGoogle from "@/contrib/google/initGoogle";
import { initToaster } from "@/utils/notify";
import { initTelemetry } from "@/background/messenger/api";
import registerMessenger from "@/extensionConsole/messenger/registration";
import { initPerformanceMonitoring } from "@/telemetry/performance";

function init(): void {
  render(<App />, document.querySelector("#container"));
}

registerMessenger();
void initGoogle();
initToaster();
initTelemetry();
void initPerformanceMonitoring();
init();
