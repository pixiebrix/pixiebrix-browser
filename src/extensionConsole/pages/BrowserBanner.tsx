/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import { type RootState } from "@/store/optionsStore";
import { selectBrowserWarningDismissed } from "@/store/settings/settingsSelectors";
import Banner from "@/components/banner/Banner";
import { useDispatch, useSelector } from "react-redux";
import { Button } from "react-bootstrap";
import settingsSlice from "@/store/settings/settingsSlice";

import { isGoogleChrome } from "@/utils/browserUtils";
import useManagedStorageState from "@/store/enterprise/useManagedStorageState";

const BrowserBanner: React.VoidFunctionComponent = () => {
  const dispatch = useDispatch();

  const browserWarningDismissed = useSelector<RootState, boolean>(
    selectBrowserWarningDismissed
  );

  const enterpriseState = useManagedStorageState();

  if (
    browserWarningDismissed ||
    isGoogleChrome() ||
    enterpriseState.isLoading ||
    enterpriseState.data?.disableBrowserWarning
  ) {
    return null;
  }

  return (
    <Banner variant="warning">
      PixieBrix officially supports Google Chrome. Some functionality may not be
      available on your browser.
      <Button
        className="info ml-3"
        size="sm"
        onClick={() => {
          dispatch(settingsSlice.actions.dismissBrowserWarning());
        }}
      >
        OK
      </Button>
    </Banner>
  );
};

export default BrowserBanner;
