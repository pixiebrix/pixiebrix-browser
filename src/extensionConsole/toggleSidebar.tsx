export const SIDEBAR_ID = "sidebar";
const smallScreenMediaQuery = "(max-width: 991px)";
const sidebarActiveClassName = "active"; // Used to show/hide navbar on small screen
const sidebarIconOnlyClassName = "sidebar-icon-only"; // Used to show/hide navbar on big screen

function getSidebar(): Element {
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- We know it exists
  return document.querySelector(`#${SIDEBAR_ID}`)!;
}

export const toggleSidebar = () => {
  if (window.matchMedia(smallScreenMediaQuery).matches) {
    getSidebar().classList.toggle(sidebarActiveClassName);
  } else {
    document.body.classList.toggle(sidebarIconOnlyClassName);
  }
};

export const closeSidebarOnSmallScreen = () => {
  if (window.matchMedia(smallScreenMediaQuery).matches) {
    getSidebar().classList.remove(sidebarActiveClassName);
  }
};
