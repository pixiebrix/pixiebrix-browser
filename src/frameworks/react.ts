/*
 * Copyright (C) 2020 Pixie Brix, LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import fromPairs from "lodash/fromPairs";

interface DOMFiber {
  _instance: any;
  _currentElement: {
    _owner: DOMFiber;
  };
}

interface ComponentFiber {
  return: ComponentFiber;
  memoizedProps: { [key: string]: unknown };
}

export class ComponentNotFoundError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ComponentNotFoundError";
  }
}

export function readReactProps(
  fiber: ComponentFiber,
  rootProp?: string
): Record<string, unknown> {
  // getComponentFiber is already returning the "return" for the component
  const memoizedProps = fiber.memoizedProps;
  console.debug("Memoized props", memoizedProps);
  const props = rootProp ? memoizedProps[rootProp] : memoizedProps;
  return fromPairs(Object.entries(props).filter(([key]) => key !== "children"));
}

export function findReactComponent(
  dom: HTMLElement,
  traverseUp = 0
): ComponentFiber {
  // https://stackoverflow.com/questions/29321742/react-getting-a-component-from-a-dom-element-for-debugging
  // https://stackoverflow.com/a/39165137/402560

  const key = Object.keys(dom).find((key) =>
    key.startsWith("__reactInternalInstance$")
  );
  // @ts-ignore: works in practice
  const domFiber: DOMFiber | null = dom[key];

  if (domFiber == null) {
    throw new ComponentNotFoundError("React fiber not found at selector");
  }

  // react <16
  // TODO: find a react 15.0 app to test this on
  if (domFiber._currentElement) {
    throw new Error("Support for React <16 not implemented");
    // let componentFiber = domFiber._currentElement._owner;
    // for (let i = 0; i < traverseUp; i++) {
    //     componentFiber = componentFiber._currentElement._owner;
    // }
    // return componentFiber._instance;
  }

  // react 16+
  const getComponentFiber = (fiber: any) => {
    //return fiber._debugOwner; // this also works, but is __DEV__ only
    let parentFiber = fiber.return;
    while (typeof parentFiber.type == "string") {
      parentFiber = parentFiber.return;
    }
    return parentFiber;
  };
  let componentFiber = getComponentFiber(domFiber);
  for (let i = 0; i < traverseUp; i++) {
    componentFiber = getComponentFiber(componentFiber);
  }
  return componentFiber;
}
