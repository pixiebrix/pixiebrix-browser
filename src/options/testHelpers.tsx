/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { configureStore } from "@reduxjs/toolkit";
import { authSlice } from "@/auth/authSlice";
import extensionsSlice from "@/store/extensionsSlice";
import settingsSlice from "@/store/settingsSlice";
import { blueprintModalsSlice } from "@/options/pages/blueprints/modals/blueprintModalsSlice";
import { createRenderWithWrappers } from "@/testUtils/testHelpers";
import blueprintsSlice from "@/options/pages/blueprints/blueprintsSlice";
import { recipesSlice } from "@/recipes/recipesSlice";

const renderWithWrappers = createRenderWithWrappers(() =>
  configureStore({
    reducer: {
      auth: authSlice.reducer,
      settings: settingsSlice.reducer,
      options: extensionsSlice.reducer,
      blueprintModals: blueprintModalsSlice.reducer,
      blueprints: blueprintsSlice.reducer,
      recipes: recipesSlice.reducer,
    },
  })
);

// eslint-disable-next-line import/export -- re-export RTL
export * from "@testing-library/react";
// eslint-disable-next-line import/export -- override render
export { renderWithWrappers as render };
