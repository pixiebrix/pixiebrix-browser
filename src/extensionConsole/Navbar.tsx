/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import styles from "./Navbar.module.scss";

import React, { useEffect } from "react";
import { Nav } from "react-bootstrap";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faBars, faExternalLinkAlt } from "@fortawesome/free-solid-svg-icons";
import { Link } from "react-router-dom";
import { getBaseURL } from "@/services/baseService";
import { useAsyncState } from "@/hooks/common";
import {
  addListener as addAuthListener,
  readPartnerAuthData,
  removeListener as removeAuthListener,
} from "@/auth/token";
import { useSelector } from "react-redux";
import { toggleSidebar } from "./toggleSidebar";
import { type SettingsState } from "@/store/settings/settingsTypes";
import cx from "classnames";
import { selectAuth } from "@/auth/authSelectors";
import { type ThemeLogo } from "@/themes/themeUtils";
import useLinkState from "@/auth/useLinkState";
import { DEFAULT_SERVICE_URL } from "@/urlConstants";

function useAdminConsoleUrl(): string {
  // Need to update serviceURL on changes to partner auth data:
  // https://github.com/pixiebrix/pixiebrix-extension/issues/4594
  const [serviceURL, _serviceUrlPending, _serviceUrlError, refreshServiceUrl] =
    useAsyncState<string>(async () => {
      const baseURL = await getBaseURL();
      const partnerAuth = await readPartnerAuthData();
      const url = partnerAuth?.token
        ? new URL("partner-auth", baseURL)
        : new URL(baseURL);
      return url.toString();
    });

  useEffect(() => {
    // Listen for token invalidation
    const handler = async () => {
      void refreshServiceUrl();
    };

    addAuthListener(handler);

    return () => {
      removeAuthListener(handler);
    };
  }, [refreshServiceUrl]);

  return serviceURL ?? DEFAULT_SERVICE_URL;
}

const Navbar: React.FunctionComponent<{ logo: ThemeLogo }> = ({ logo }) => {
  const { email } = useSelector(selectAuth);

  const { hasToken: connected, tokenLoading: connectedPending } =
    useLinkState();

  const serviceURL = useAdminConsoleUrl();

  const mode = useSelector<{ settings: SettingsState }, string>(
    ({ settings }) => settings.mode
  );

  // Use `connectedPending` to optimistically show the toggle
  const showNavbarToggle = mode === "local" || connected || connectedPending;

  return (
    <nav className="navbar default-layout-navbar col-lg-12 col-12 p-0 fixed-top d-flex flex-row">
      <div className={cx(styles.collapsedWrapper, "navbar-brand-wrapper")}>
        <Link className="navbar-brand brand-logo" to="/">
          <img src={logo.regular} alt="PixieBrix logo" />
        </Link>
        <Link className="navbar-brand brand-logo-mini" to="/">
          <img src={logo.small} alt="PixieBrix mini logo" />
        </Link>
        {showNavbarToggle && (
          <button
            className={cx("navbar-toggler", styles.collapsedSidebarToggler)}
            type="button"
            onClick={toggleSidebar}
          >
            <FontAwesomeIcon icon={faBars} />
          </button>
        )}
      </div>
      <div className={cx(styles.expandedWrapper, "navbar-menu-wrapper")}>
        {showNavbarToggle && (
          <button
            className="navbar-toggler align-self-center"
            type="button"
            onClick={toggleSidebar}
          >
            <FontAwesomeIcon icon={faBars} />
          </button>
        )}

        <div
          className={cx(styles.extensionConsoleHeading, "d-none d-md-inline")}
        >
          Extension Console
        </div>

        <ul className="navbar-nav navbar-nav-right flex-grow-1 justify-content-end">
          {
            <Nav.Link className="px-3" target="_blank" href={serviceURL}>
              <FontAwesomeIcon icon={faExternalLinkAlt} className="mr-1" />
              Open Admin Console
            </Nav.Link>
          }
          {connected && email && <div className="text-black">{email}</div>}
        </ul>
      </div>
    </nav>
  );
};

export default Navbar;
