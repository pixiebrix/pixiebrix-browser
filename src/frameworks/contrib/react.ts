/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { pickBy } from "lodash";
import { ComponentNotFoundError, ignoreNotFound } from "@/frameworks/errors";
import { RootInstanceVisitor } from "@/frameworks/scanner";
import { ReadableComponentAdapter, traverse } from "@/frameworks/component";
import { isNode } from "@/frameworks/dom";

// React architecture references:
// https://github.com/acdlite/react-fiber-architecture
// https://indepth.dev/posts/1008/inside-fiber-in-depth-overview-of-the-new-reconciliation-algorithm-in-react#memoizedprops
// https://github.com/Venryx/mobx-devtools-advanced/blob/master/Docs/TreeTraversal.md
// https://stackoverflow.com/questions/29321742/react-getting-a-component-from-a-dom-element-for-debugging

interface RootInstance {
  _reactRootContainer: unknown;
}

/**
 * Structure from:
 * - https://github.com/facebook/react/blob/50d9451f320a9aaf94304209193562cc385567d8/packages/react-devtools-shared/src/backend/legacy/renderer.js#L739
 * - https://github.com/facebook/react/blob/v15.0.0/src/renderers/shared/reconciler/ReactReconciler.js
 */
interface LegacyInstance {
  /**
   * Public instance
   */
  _instance: unknown;

  _currentElement: {
    _owner: LegacyInstance;
  };
}

/**
 * Type definition from: https://github.com/facebook/react/blob/4c6470cb3b821f3664955290cd4c4c7ac0de733a/packages/react-reconciler/src/ReactInternalTypes.js#L62
 */
interface Fiber {
  /**
   * The Fiber to return to after finishing processing this one.
   * This is effectively the parent, but there can be multiple parents (two)
   * so this is only the parent of the thing we're currently processing.
   * It is conceptually the same as the return address of a stack frame.
   */
  return: Fiber | null;

  /**
   * The props used to create the output.
   */
  memoizedProps: { [prop: string]: unknown };

  /**
   * The resolved function/class associated with this fiber.
   */
  type: string | object;

  /**
   * The local state associated with this fiber.
   */
  stateNode: Node | object;
}

export function isManaged(node: Node): boolean {
  return Object.keys(node).some((key) =>
    key.startsWith("__reactInternalInstance$")
  );
}

export function hasReactProps(fiber: Fiber): boolean {
  return Object.keys(fiber.memoizedProps).some((x) => x !== "children");
}

export function readReactProps(fiber: Fiber): { [prop: string]: unknown } {
  return pickBy(fiber.memoizedProps, (value, key) => key !== "children");
}

function getComponentFiber(fiber: Fiber): Fiber {
  // return fiber._debugOwner; // this also works, but is __DEV__ only
  let parentFiber = fiber.return;
  while (typeof parentFiber.type === "string") {
    // string for HTML nodes, so traverse
    parentFiber = parentFiber.return;
  }
  return parentFiber;
}

export function findReactComponent(node: Node, traverseUp = 0): Fiber {
  // https://stackoverflow.com/a/39165137/402560
  const key = Object.keys(node).find((key) =>
    key.startsWith("__reactInternalInstance$")
  );

  const domFiber: Fiber | LegacyInstance | null = (node as any)[key];

  if (domFiber == null) {
    throw new ComponentNotFoundError("React fiber not found for element");
  }

  if ("_currentElement" in domFiber) {
    // FIXME: test this for React <16
    const owner = (x: LegacyInstance) => x._currentElement._owner;
    const fiber = traverse(owner, owner(domFiber), traverseUp);
    return fiber._instance as Fiber;
  } else {
    return traverse(getComponentFiber, getComponentFiber(domFiber), traverseUp);
  }
}

export class ReactRootVisitor implements RootInstanceVisitor<RootInstance> {
  public rootInstances: RootInstance[] = [];
  visit(node: Element | Node): boolean {
    if ("_reactRootContainer" in node) {
      this.rootInstances.push(node);
      return false;
    }
    return true;
  }
}

export const adapter: ReadableComponentAdapter<Fiber> = {
  isManaged,
  getComponent: (node) => ignoreNotFound(() => findReactComponent(node, 0)),
  getParent: getComponentFiber,
  getNode: (instance) =>
    isNode(instance.stateNode) ? instance.stateNode : null,
  getData: readReactProps,
  hasData: hasReactProps,
};

export default adapter;
