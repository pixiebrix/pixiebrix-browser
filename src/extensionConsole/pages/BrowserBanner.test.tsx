/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import { render } from "@/extensionConsole/testHelpers";
import BrowserBanner from "@/extensionConsole/pages/BrowserBanner";
import { waitForEffect } from "@/testUtils/testHelpers";
import { screen } from "@testing-library/react";
import { INTERNAL_reset } from "@/store/enterprise/managedStorage";

beforeEach(async () => {
  // eslint-disable-next-line new-cap -- test helper method
  INTERNAL_reset();
  await browser.storage.managed.clear();
});

describe("BrowserBanner", () => {
  it("renders warning for jest environment", async () => {
    await browser.storage.managed.set({ disableBrowserWarning: false });

    const wrapper = render(
      <div>
        <BrowserBanner />
      </div>
    );

    await waitForEffect();

    expect(
      screen.queryByText("PixieBrix officially supports Google Chrome", {
        exact: false,
      })
    ).toBeInTheDocument();

    expect(wrapper.asFragment()).toMatchSnapshot();
  });

  it("suppress with managed storage", async () => {
    await browser.storage.managed.set({ disableBrowserWarning: true });

    const wrapper = render(
      <div>
        <BrowserBanner />
      </div>
    );

    await waitForEffect();

    expect(
      screen.queryByText("PixieBrix officially supports Google Chrome.", {
        exact: false,
      })
    ).not.toBeInTheDocument();

    expect(wrapper.asFragment()).toMatchSnapshot();
  });
});
