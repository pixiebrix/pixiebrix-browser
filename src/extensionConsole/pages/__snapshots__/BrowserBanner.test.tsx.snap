// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrowserBanner renders warning for jest environment 1`] = `
<DocumentFragment>
  <div>
    <div
      class="root warning"
    >
      <div
        class="mx-auto d-flex"
      >
        <div
          class="flex-grow-1"
        />
        <div
          class="align-self-center"
        >
          PixieBrix officially supports Google Chrome. Some functionality may not be available on your browser.
          <button
            class="info ml-3 btn btn-primary btn-sm"
            type="button"
          >
            OK
          </button>
        </div>
        <div
          class="flex-grow-1"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BrowserBanner suppress with managed storage 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;
